src/components/Layer.js

```diff
@@ -1,756 +1,461 @@
-import {confirmAlert} from 'react-confirm-alert'; // Import
-import neo4jService from '../services/neo4jService'
-import {ColorPicker} from "./ColorPicker";
-
-
-import 'react-confirm-alert/src/react-confirm-alert.css'; // Import css
+import Accordion from 'react-bootstrap/Accordion';
+import { Button, Form } from 'react-bootstrap';
+import { confirmAlert } from 'react-confirm-alert';
+import { neo4jService } from '../services/neo4jService'
+import {
+  LayerTypeForm,
+  MapRenderingForm,
+  LatLonLayerForm,
+  PointLayerForm,
+  SpatialLayerForm,
+  CypherLayerForm
+} from './Layer.forms';
+
+// Import css
+import 'react-confirm-alert/src/react-confirm-alert.css';
 // css needed for CypherEditor
 import "codemirror/lib/codemirror.css";
 import "codemirror/addon/lint/lint.css";
 import "codemirror/addon/hint/show-hint.css";
 import "cypher-codemirror/dist/cypher-codemirror-syntax.css";
 
 import {
-	NEW_LAYER,
-	LAYER_TYPE_LATLON,
-	LAYER_TYPE_POINT,
-	LAYER_TYPE_CYPHER,
-	LAYER_TYPE_SPATIAL,
-	RENDERING_MARKERS,
-	RENDERING_POLYLINE,
-	RENDERING_HEATMAP,
-	RENDERING_CLUSTERS
+  NEW_LAYER,
+  LAYER_TYPE_LATLON,
+  LAYER_TYPE_POINT,
+  LAYER_TYPE_CYPHER,
+  LAYER_TYPE_SPATIAL,
+  RENDERING_HEATMAP,
 } from './constants';
 
-export class Layer extends Component {
-
-	constructor(props) {
-		super(props);
-
-		this.state = props.layer;
-
-		this.driver = props.driver;
-
-		this.showQuery = this.showQuery.bind(this);
-		this.handleNameChange = this.handleNameChange.bind(this);
-		this.handleLayerTypeChange = this.handleLayerTypeChange.bind(this);
-		this.handleNodeLabelChange = this.handleNodeLabelChange.bind(this);
-		this.handleLatPropertyChange = this.handleLatPropertyChange.bind(this);
-		this.handleLonPropertyChange = this.handleLonPropertyChange.bind(this);
-		this.handlePointPropertyChange = this.handlePointPropertyChange.bind(this);
-		this.handleTooltipPropertyChange = this.handleTooltipPropertyChange.bind(this);
-		this.handleLimitChange = this.handleLimitChange.bind(this);
-		this.handleColorChange = this.handleColorChange.bind(this);
-		this.handleRenderingChange = this.handleRenderingChange.bind(this);
-		this.handleRadiusChange = this.handleRadiusChange.bind(this);
-		this.handleCypherChange = this.handleCypherChange.bind(this);
-		this.handleSpatialLayerChanged = this.handleSpatialLayerChanged.bind(this);
-
-	};
-
 
-	componentDidMount() {
-		// list of available nodes
-		this.getNodes();
-		this.getPropertyNames();
-		this.hasSpatialPlugin();
-		this.getSpatialLayers();
-	}
-
-
-	updateBounds = () => {
-		/* Compute the map bounds based on `this.state.data`
+export const Layer = React.memo(({ layer, addLayer, updateLayer, removeLayer }) => {
+
+  const [state, setState] = React.useState(layer);
+
+  const updateState = (update, callback = undefined) => {
+    setState({ ...state, ...update }, callback);
+  }
+
+  React.useEffect(() => {
+
+    getNodes();
+    getPropertyNames();
+    hasSpatialPlugin();
+    getSpatialLayers();
```

Comment:
Note for later: the getSpatialLayers function must be called ONLY IF hasSpatialPlugin is true, otherwise it will always fail, and never be used anyway.

Reply:
Sounds like a good TODO comment. I'll add one

Reply:
Followed this thread and improved the logic flow for how we call getSpatialLayers. Now getSpatialLayers does and internal check for hasSpatial which is deprecated as a top level neo4jService api.
There's still some points of clarification to be made, like why we store _spatialLayers_ (array) and _spatialLayer_ in Layer definition we pass around. Marked with TODO